#### NLTK: natural language ?? (library)
#### python은 2.7 ver-> 3.0
#### change가 많아서 호환이 잘 안됨
#### python 사용자 중 2.7에 머물기도 해
#### re: regular expression 관련 패키지
#### function 사용된 거: re.search, re.findall 이 정도만 사용한 상태
#### bracket 해서 list comprehension에 익숙해질 것
#### 코드창인데 markdown처럼 comment를 넣으려면 앞에 # 넣어도 ok
#### range로 #을 넣는 기능을 위해서(너무 계속 # 붙이려면 귀찮으니까)  '''~''' or """~""" (적어놓았으나 실행되진 않음)

In [4]:
nltk.download()

NameError: name 'nltk' is not defined

In [2]:
import nltk
import re

In [8]:
var="""
Basic Regular Expression Meta-Characters, Including Wildcards, Ranges and Closures
.	        Wildcard, matches any character
^abc	    Matches some pattern abc at the start of a string
abc$	    Matches some pattern abc at the end of a string
[abc]	    Matches one of a set of characters
[^abc]      Matches anything but a set of characters
[A-Z0-9]	Matches one of a range of characters
ed|ing|s	Matches one of the specified strings (disjunction)
*	        Zero or more of previous item, e.g. a*, [a-z]* (also known as Kleene Closure)
+	        One or more of previous item, e.g. a+, [a-z]+
?	        Zero or one of the previous item (i.e. optional), e.g. a?, [a-z]?
{n}	        Exactly n repeats where n is a non-negative integer
{n,}	    At least n repeats
{,n}	    No more than n repeats
{m,n}	    At least m and no more than n repeats
a(b|c)+	    Parentheses that indicate the scope of the operators

"""

#### 명심: string에 적용되는 함수이다
#### string에 적용되는 것이 regexp
#### 사전을 통째로  nltk에서 가져옴

In [13]:
dic = nltk.corpus.words.words('en')
print(type(dic))

<class 'list'>


In [14]:
a=[]
for w in nltk.corpus.words.words('en'):
    if w.islower():
        a.append(w)

#### in 뒤의 list
#### 그 리스트 안의 단어가 소문자일 때만 wordlist 안의 리스트에 쌓인다.
#### 최초 10개만 보여다오

In [15]:
wordlist = [w for w in dic if w.islower()]
print(wordlist[:10])

['a', 'aa', 'aal', 'aalii', 'aam', 'aardvark', 'aardwolf', 'aba', 'abac', 'abaca']


In [33]:
wordlist = [w for w in nltk.corpus.words.words('en') if w.islower()]
print(wordlist[:10])

['a', 'aa', 'aal', 'aalii', 'aam', 'aardvark', 'aardwolf', 'aba', 'abac', 'abaca']


In [31]:
a==wordlist

True

#### regexp -string에 적용된다, list에 적용되지 않는다
#### for loop의 의미: 어떤 사전이 있는데 이 조건에 해당하는 단어를 찾아라(X)
#### regexp를 리스트 전체에 적용시키기 위해서는 한 요소씩(하나하나씩) 적용해야 한다(하나하나는 string이니까)

#### 첫째: 찾는 조건, 둘째: 찾는 대상(리스트도 될 수 있고, string도 들어갈 수 있다)
#### 뒤어세 앞의 내용을 찾아라
#### 보여주는 방식: 찾은 걸 보여줄 수 있고, 찾은 range를 보여줄 수도 있다
#### 여기선 둘 다 나옴

In [19]:
re.search('abc','abcde')   # 함수의 입력 갯수: 2개
print(type(re.search('abc','abcde')))

<class 're.Match'>


#### output으로 아무것도 없는 이유는 2째 input에 1째 input에 해당하는 부분이 없으니까

In [18]:
re.search('ed','abcde')

In [17]:
x = re.search('abc','abcde')
print(x)

<re.Match object; span=(0, 3), match='abc'>


#### wordlist 안에 있는 단어들이 w로 들어가서 for loop을 돈다(횟수: len(wordlist))
#### w에 쌓일지 말지는 if문에서 결정됨 -ed로 끝나는 단어 가져오기, ed로 끝나지 않으면 그냥 패스하기
#### 그렇게 해서 만들어진 리스트 중 처음부터 10번째 것까지만 result에 들어간다.
#### 이럴 때 if 문이 없다면([:10]을 입력하지 않았다고 가정하자) result == wordlist

In [18]:
result = [w for w in wordlist if re.search('ed$', w)][:10]
print(result)
result = [w for w in wordlist if re.search('ed$', w)]
print(result[:10])

['abaissed', 'abandoned', 'abased', 'abashed', 'abatised', 'abed', 'aborted', 'abridged', 'abscessed', 'absconded']
['abaissed', 'abandoned', 'abased', 'abashed', 'abatised', 'abed', 'aborted', 'abridged', 'abscessed', 'absconded']


In [ ]:
result = [w for w in wordlist if re.search('ed$', w)][:10]
print(result[:10])  # 여기서 10은 redundant

#### ^: 앞, $: 뒤
#### 3번째 character가 j, 6번째 character가 t, 그리고 t 뒤에 2개의 character가 있는 단어를 찾아라(각각의 .== 각각 하나의 character)

In [21]:
result = [w for w in wordlist if re.search('^..j..t..$', w)]    # 출력되는 리스트의 단어릐 length: 8
print(result[:10])
print(result)
print(len(result))

['abjectly', 'adjuster', 'dejected', 'dejectly', 'injector', 'majestic', 'objectee', 'objector', 'rejecter', 'rejector']
['abjectly', 'adjuster', 'dejected', 'dejectly', 'injector', 'majestic', 'objectee', 'objector', 'rejecter', 'rejector', 'unjilted', 'unjolted', 'unjustly']
13


In [47]:
result = [w for w in wordlist if re.search('^..j..t..$', w)][:10]    # 출력되는 리스트의 단어릐 length: 8
print(result[:10])
print(type(result))

['abjectly', 'adjuster', 'dejected', 'dejectly', 'injector', 'majestic', 'objectee', 'objector', 'rejecter', 'rejector']
<class 'list'>


In [48]:
result = [w for w in wordlist if re.search('^..j....t..$', w)][:10]
print(result[:10])

['adjacently', 'objurgator', 'unjacketed', 'unjesuited']


#### 첫째 자리: g,h,i 중 하나
#### 둘째 자리: m,n,o 중 하나
#### 셋째 자리: j,l,k 중 하나
#### 넷째(=마지막자리($이 입력되어 있으므로)) 자리: d,e,f 중 하나
#### 인 단어를 찾아서 앞의 10개까지만 result에 넣어라

In [7]:
result = [w for w in wordlist if re.search('^[ghi][mno][jlk][def]$', w)][:10]
print(result[:10])

['gold', 'golf', 'hold', 'hole']


#### 첫째 자리: a 또는 h
#### 둘째 자리~끝 자리: 아무것도 오지 않을 수 있고, 온다면 a또는 h가 온다.
#### +의 의미: 1회 이상
#### 10개까지 print하라고 했으나 원래 변수의 길이가 10보다 짧으면 짧은 대로 print된다.

In [8]:
result = [w for w in wordlist if re.search('^[ah]+$', w)][:10]
print(result[:10])

['a', 'aa', 'ah', 'aha', 'h', 'ha', 'hah']


#### 너무 많아서 끝 부분이 ...으로 나옴:  ('3.80', ...])
#### world street journal
#### wsj에서 나온 단어들이 나온 것
#### type(wsj) == list

In [4]:
wsj = sorted(set(nltk.corpus.treebank.words()))
# wsj

#### 처음(^이 있으므로): 0~9 중 하나(+: 1회 이상 올 수 있다)
### \ . :  특수문자인 점을 print하는 방법(실제 .을 찾는 방법)
#### 끝(달러모양): 0~9 중 하나(+: 1회 이상 올 수 있다)
#### 만약 [09] -> 0 또는 9
### \ + : 특수문자인 +을 찾아라
#### 실수를 찾는 셈(음수 제외)
#### '^'와 '$' 없애면 정해진 패턴이 아무데서나 포함되기만 하면 됨

In [25]:
wordlist = [w for w in wsj if re.search('^[0-9]+\.[0-9]+$', w)]
print(wordlist[:10])
print(wordlist[470:])

['0.0085', '0.05', '0.1', '0.16', '0.2', '0.25', '0.28', '0.3', '0.4', '0.5']
['9.9', '92.9', '93.3', '93.9', '94.2', '94.8', '95.09', '96.4', '98.3', '99.1', '99.3']


#### 알파벳이 뭐가 나오든!
#### 알파벳이든 숫자든 자리 1개 = .
#### 왜 달러모양이 2개? '\달러모양'의 의미: 특수문자인 달러모양기호가 붙어있다
#### 알파벳 아무거나 + $로 끝나는 것을 찾아라
#### [A-Z0-9]: 알파벳 또는 숫자 1개 나옴.(특수문자를 제외한 것) len = 1
#### [^abc] : a,b,c가 아닌 것 아무거나 중 1개만 나와라

In [27]:
wordlist = [w for w in wsj if re.search('^[A-Z]+\$$', w)]
print(wordlist[:10])

['C$', 'US$']


#### 리스트가 비어 있다:1개이상의 소문자 뒤에 달러모양이 붙은 경우는 없다

In [28]:
wordlist = [w for w in wsj if re.search('^[a-z]+\$$', w)]
print(wordlist[:10])

[]


#### 결과의 의미: 알파벳이 1개 이상 나오고 뒤에 달러모양이 오는 경우는 대문자 1~2개 뒤에 달러모양이 오는 경우만 있다.

In [29]:
wordlist = [w for w in wsj if re.search('^[A-Za-z]+\$$', w)]
print(wordlist[:10])

['C$', 'US$']


#### {숫자} : 반복 횟수
#### 0~9에서 하나의 숫자의 골라 나오는 과정이 4번 반복된다.

In [12]:
wordlist = [w for w in wsj if re.search('^[0-9]{4}$', w)]
print(wordlist[:10])

['1614', '1637', '1787', '1901', '1903', '1917', '1925', '1929', '1933', '1934']


#### 리스트 요소의 특징
#### 처음에 숫자가 하나 이상 나온다
#### -: 그냥 입력되고 출력된다(그냥 특수기호일 뿐이라 별다른 의미가 없다)
#### 끝엔 알파벳이 3~5자리가 나온다
#### 하나 이상이라고 지정했는데 하나짜리는 없나봐(숫자)

In [13]:
wordlist = [w for w in wsj if re.search('^[0-9]+-[a-z]{3,5}$', w)]
print(wordlist[:10])

['10-day', '10-lap', '10-year', '100-share', '12-point', '12-year', '14-hour', '15-day', '150-point', '190-point']


#### <이 리스트 요소의 특징>
#### 1) 알파벳이 5개 이상
#### 2) -가 쓰인다
#### 3) 알파벳이 2~3개
#### 4) -가 쓰임
#### 5) 알파벳이 6개 이하인 것
#### 하나의 결과물을 위해서 다양한 regexp을 쓸 수 있다

In [14]:
wordlist = [w for w in wsj if re.search('^[a-z]{5,}-[a-z]{2,3}-[a-z]{,6}$', w)]
print(wordlist[:10])

['black-and-white', 'bread-and-butter', 'father-in-law', 'machine-gun-toting', 'savings-and-loan']


#### | : '또는'
#### ed로 끝나거나 ing로 끝나는 것
#### (||||) 이렇게 또는의 대상이 되는 것들은 소괄호로 묶어야지

In [15]:
wordlist = [w for w in wsj if re.search('(ed|ing)$', w)]
print(wordlist[:10])

['62%-owned', 'Absorbed', 'According', 'Adopting', 'Advanced', 'Advancing', 'Alfred', 'Allied', 'Annualized', 'Anything']


#### 또는의 대상이 되는 string의 길이가 1일 수 있다.

In [7]:
wordlist = [w for w in wsj if re.search('(e|i)$', w)]
print(wordlist[:10])

["'re", "'ve", '100-megabyte', '100-share', '120-a-share', '18-a-share', '240-page', '30-minute', '30-share', '300-a-share']


#### nltk.FreqDist(vs for word in wsj) 이걸 먼저 생각해라
#### :wsj에 있는 단어들을 word에 넣는데, 이때 각 요소가 몇 번 나오는지를 짝 지어 줘라
#### FreqDist의 결과물은 list가 아니다

In [12]:
fdist = nltk.FreqDist(word for word in wsj)
print(fdist)
print(type(fdist))
print(len(fdist))
print(len(wsj))

<FreqDist with 12408 samples and 12408 outcomes>
<class 'nltk.probability.FreqDist'>
12408
12408


#### re.search: 뭐가 있느냐 없느냐를 찾아줌
#### re.findall: 2째input에 1째input에 해당하는 것이 있으면 그냥 다 찾아서 리스트로 만들어서 줘라
#### re.search와 re.findall이 주는 output의 type은 다르다

In [22]:
print(re.findall('ab','ababdddabkabddd'))
print(type(re.findall('ab','ababdddabkabddd')))

['ab', 'ab', 'ab', 'ab']
<class 'list'>


#### re.Match에는 len 함수를 적용할 수 없다.

In [24]:
print(re.search('ab','ababdddabkabddd'))
print(type(re.search('ab','ababdddabkabddd')))
print(len(re.search('ab','ababdddabkabddd')))

<re.Match object; span=(0, 2), match='ab'>
<class 're.Match'>


TypeError: object of type 're.Match' has no len()

#### 왜 하는가? - 이 사전에서 모음을 다 뽑아와라
#### 하나 이상이 되는 것을 뽑아오라고 할 수도 있다

In [8]:
atleastone1 = [vs for word in wsj for vs in re.findall(r'[aeiou]{1,}',word)]
print(type(atleastone1))

<class 'list'>


In [9]:
atleastone2 = [vs for word in wsj for vs in re.findall(r'[aeiou]+',word)]

In [10]:
atleastone1 == atleastone2

True

In [11]:
print(len(atleastone1))

26119


In [12]:
vs for word in wsj for vs in re.findall(r'[aeiou]{2,}',word)

SyntaxError: invalid syntax (<ipython-input-12-a241399327fd>, line 1)

In [33]:
practice = [vs for word in wsj for vs in re.findall(r'[aeiou]{2,}',word)]  # type: list
print(len(practice))
print(practice)

3405
['ea', 'oi', 'ea', 'ou', 'oi', 'ea', 'ea', 'oi', 'oi', 'ea', 'io', 'ea', 'ea', 'ea', 'oi', 'ea', 'ea', 'ea', 'ea', 'ea', 'ea', 'ea', 'ee', 'ea', 'ea', 'ea', 'ea', 'ea', 'ea', 'ea', 'ea', 'oi', 'ea', 'ea', 'ou', 'ou', 'ou', 'ie', 'ui', 'io', 'ua', 'io', 'ai', 'ai', 'ai', 'io', 'ie', 'ue', 'ue', 'ia', 'ie', 'ea', 'ai', 'ou', 'ia', 'ei', 'ie', 'ea', 'ea', 'ie', 'ia', 'ia', 'ua', 'ie', 'io', 'ea', 'ia', 'io', 'ui', 'ia', 'ia', 'ea', 'iai', 'ai', 'ia', 'ia', 'ia', 'ia', 'ia', 'io', 'oo', 'io', 'ia', 'ia', 'ia', 'ia', 'ue', 'ea', 'ai', 'ai', 'ue', 'ie', 'au', 'ea', 'ea', 'ea', 'ea', 'eau', 'au', 'ei', 'ei', 'ei', 'ei', 'ei', 'ia', 'ie', 'io', 'ue', 'oa', 'oei', 'oe', 'ia', 'oo', 'oo', 'oo', 'eau', 'ou', 'ou', 'ai', 'ou', 'ai', 'oo', 'ea', 'au', 'ia', 'ea', 'ea', 'ee', 'ia', 'ai', 'oa', 'oo', 'oo', 'oo', 'ei', 'ei', 'ea', 'ui', 'ui', 'eau', 'ie', 'ia', 'ia', 'ai', 'eau', 'ia', 'ea', 'ie', 'ie', 'oo', 'ai', 'ai', 'io', 'ie', 'ie', 'eau', 'ee', 'ee', 'ea', 'io', 'oo', 'ia', 'ie', 'ui', 'io

#### wsj의 첫째 요소를 word에 넣는다
#### word에서 r'[aeiou]{2,}'을 만족시키는 부분을 모두 찾아 만든 list의 요소를 아래 vs에 하나 넣는다.
#### 아래 for loop을 통해 vs에 add up해서 다 넣고나면 vs가 하나의 list로 완성되고
#### 그 완성된 list가 위vs(최종 list)에 추가되고, 다음 순서(wsj의 둘째 요소)의 for loop을 돈다.

#### vs for word in wsj for vs in re.findall(r'[aeiou]{2,}',word)이 대괄호로 싸여 있어야 하는데 생략됨.
#### FrqDist의 input의 type: list
#### 아무생각 없이 모여있는 정보(3405개의 정보, 43가지 종류)에서 가장 많이 언급된 순서대로 정보들을 나열한 list를 보여줘: most_common
#### most_common에 길이를 지정하면 list, 지정하지 않으면(그냥 most_common을 쓰면) method

In [5]:
fdist = nltk.FreqDist(vs for word in wsj for vs in re.findall(r'[aeiou]{2,}',word))
print(fdist)
print(fdist.most_common(10))
print(fdist.most_common)

<FreqDist with 43 samples and 3405 outcomes>
[('io', 549), ('ea', 476), ('ie', 331), ('ou', 329), ('ai', 261), ('ia', 253), ('ee', 217), ('oo', 174), ('ua', 109), ('au', 106)]
<bound method Counter.most_common of FreqDist({'io': 549, 'ea': 476, 'ie': 331, 'ou': 329, 'ai': 261, 'ia': 253, 'ee': 217, 'oo': 174, 'ua': 109, 'au': 106, ...})>


In [13]:
print(type(fdist.most_common(10)))
print(type(fdist.most_common))

<class 'list'>
<class 'method'>


In [16]:
word = 'supercalifragilisticexpialidocious'

In [ ]:
#### 모음 찾아라
#### 모음 중 가장 많이 출현하는 것은? -해봐!

In [17]:
result = re.findall(r'[aeiou]', word)
print(result)

['u', 'e', 'a', 'i', 'a', 'i', 'i', 'i', 'e', 'i', 'a', 'i', 'o', 'i', 'o', 'u']


In [18]:
result = re.findall('[aeiou](..)[aeiou]', word)
print(result)

['rc', 'fr', 'st', 'xp', 'ci']


In [19]:
result = re.findall('[^aeiou].+[^aeiou]', word)
print(result)

['supercalifragilisticexpialidocious']


In [20]:
result = re.findall('[^aeiou].+?[^aeiou]', word)
print(result)

['sup', 'rcal', 'frag', 'lis', 'tic', 'xpial', 'doc']


In [21]:
wsj = sorted(set(nltk.corpus.treebank.words()))
fdist = nltk.FreqDist(vs for word in wsj for vs in re.findall(r'[aeiou]{2,}', word))
fdist.most_common(10)

[('io', 549),
 ('ea', 476),
 ('ie', 331),
 ('ou', 329),
 ('ai', 261),
 ('ia', 253),
 ('ee', 217),
 ('oo', 174),
 ('ua', 109),
 ('au', 106)]

In [22]:
regexp = r'^[AEIOUaeiou]+|[AEIOUaeiou]+$|[^AEIOUaeiou]'
def compress(word):
    pieces = re.findall(regexp, word)
    return ''.join(pieces)

english_udhr = nltk.corpus.udhr.words('English-Latin1')
print(nltk.tokenwrap(compress(w) for w in english_udhr[:75]))


Unvrsl Dclrtn of Hmn Rghts Prmble Whrs rcgntn of the inhrnt dgnty and
of the eql and inlnble rghts of all mmbrs of the hmn fmly is the fndtn
of frdm , jstce and pce in the wrld , Whrs dsrgrd and cntmpt fr hmn
rghts hve rsltd in brbrs acts whch hve outrgd the cnscnce of mnknd ,
and the advnt of a wrld in whch hmn bngs shll enjy frdm of spch and


In [23]:
rotokas_words = nltk.corpus.toolbox.words('rotokas.dic')
cvs = [cv for w in rotokas_words for cv in re.findall(r'[ptksvr][aeiou]', w)]
cfd = nltk.ConditionalFreqDist(cvs) # CFD works pairwise only
cfd.tabulate()

    a   e   i   o   u 
k 418 148  94 420 173 
p  83  31 105  34  51 
r 187  63  84  89  79 
s   0   0 100   2   1 
t  47   8   0 148  37 
v  93  27 105  48  49 


In [24]:
cv_word_pairs = [(cv, w) for w in rotokas_words for cv in re.findall(r'[ptksvr][aeiou]', w)]
cv_index = nltk.Index(cv_word_pairs)
cv_index['po']

['kaapo',
 'kaapopato',
 'kaipori',
 'kaiporipie',
 'kaiporivira',
 'kapo',
 'kapoa',
 'kapokao',
 'kapokapo',
 'kapokapo',
 'kapokapoa',
 'kapokapoa',
 'kapokapora',
 'kapokapora',
 'kapokaporo',
 'kapokaporo',
 'kapokari',
 'kapokarito',
 'kapokoa',
 'kapoo',
 'kapooto',
 'kapoovira',
 'kapopaa',
 'kaporo',
 'kaporo',
 'kaporopa',
 'kaporoto',
 'kapoto',
 'karokaropo',
 'karopo',
 'kepo',
 'kepoi',
 'keposi',
 'kepoto']

In [36]:
A = ['a','b']
B = ['c','d']
A+B
# [A,B]

['a', 'b', 'c', 'd']